contributors/design-proposals/daemonset-update.md

```diff
@@ -165,22 +187,38 @@ DaemonSets on the apiserver.
 
 For each pending DaemonSet updates, it will:
 
-1. Find all pods whose labels are matched by DaemonSet `.spec.selector`. 
-   - If `OwnerReference` is implemented for DaemonSets, filter out pods that
-     aren't controlled by this DaemonSet too.
+1. Find all pods controlled by DaemonSet
 1. Check `DaemonSetUpdateStrategy`:
    - If `OnDelete`: do nothing
    - If `RollingUpdate`:
-     - Find all daemon pods that belong to this DaemonSet using label selectors.
-       - Pods with "pod-template-generation" value equal to the DaemonSet's
-         `.spec.templateGeneration` are new pods, otherwise they're old pods.
-       - Note that pods without "pod-template-generation" labels (e.g. DaemonSet
-         pods created before RollingUpdate strategy is implemented) will be
-         seen as old pods.
+     - Find existing PodTemplates owned by DaemonSet, and sort them by the value
+       of `pod-template-generation` label
+     - Clean up PodTemplates based on `.spec.revisionHistoryLimit`
+       - Always keep the PodTemplate if its `pod-template-generation` label has
+         the same value of any existing pods owned by the DaemonSet
+       - Remove PodTemplates from the ones with the smallest `pod-template-generation`
+         to the highest, until `.spec.revisionHistoryLimit` is hit or no
+         PodTemplates are allowed to be removed 
+     - Create a PodTemplate based on DaemonSet
```

Comment:
This proposal doesn't differentiate rollback by update vs. by calling /rollback. 

Relabeling a pod is already included (see L221).

The reason for duplicating PodTemplates is to keep track of templateGeneration vs. PodSpec mapping. For example, let's say current templateGeneration is 5, v1 = v3 = v5 and there are v1 and v3 existing pods. If we reuse PodTemplates, we only know that v3 = v5 (or v1 = v5, depends on current label of the PodTemplate), and will have to kill v1 (or v3) pods. Alternatively, we can use revision annotation (instead of relabeling Pods and PodTemplates) to solve this, but templateGeneration + revision may be confusing since they have similar meaning (from the name).